array_api_tests/test_manipulation_functions.py

```diff
@@ -0,0 +1,369 @@
+    data=st.data(),
+)
+def test_concat(dtypes, kw, data):
+    axis = kw.get("axis", 0)
+    if axis is None:
+        shape_strat = hh.shapes()
+    else:
+        _axis = axis if axis >= 0 else abs(axis) - 1
+        shape_strat = shared_shapes(min_dims=_axis + 1).flatmap(
+            lambda s: concat_shapes(s, axis)
+        )
+    arrays = []
+    for i, dtype in enumerate(dtypes, 1):
+        x = data.draw(xps.arrays(dtype=dtype, shape=shape_strat), label=f"x{i}")
+        arrays.append(x)
+
+    out = xp.concat(arrays, **kw)
+
+    ph.assert_dtype("concat", dtypes, out.dtype)
+
+    shapes = tuple(x.shape for x in arrays)
+    axis = kw.get("axis", 0)
+    if axis is None:
+        size = sum(math.prod(s) for s in shapes)
+        shape = (size,)
+    else:
+        shape = list(shapes[0])
+        for other_shape in shapes[1:]:
+            shape[axis] += other_shape[axis]
+        shape = tuple(shape)
+    ph.assert_result_shape("concat", shapes, out.shape, shape, **kw)
+
+    if axis is None:
+        out_indices = (i for i in range(out.size))
+        for x_num, x in enumerate(arrays, 1):
+            for x_idx in ah.ndindex(x.shape):
+                out_i = next(out_indices)
+                assert_equals(
+                    "concat",
+                    f"x{x_num}[{x_idx}]",
+                    x[x_idx],
+                    f"out[{out_i}]",
+                    out[out_i],
+                    **kw,
+                )
+    else:
+        out_indices = ah.ndindex(out.shape)
+        for idx in axis_ndindex(shapes[0], _axis):
+            f_idx = ", ".join(str(i) if isinstance(i, int) else ":" for i in idx)
+            for x_num, x in enumerate(arrays, 1):
+                indexed_x = x[idx]
+                for x_idx in ah.ndindex(indexed_x.shape):
+                    out_idx = next(out_indices)
+                    assert_equals(
+                        "concat",
+                        f"x{x_num}[{f_idx}][{x_idx}]",
+                        indexed_x[x_idx],
+                        f"out[{out_idx}]",
+                        out[out_idx],
+                        **kw,
+                    )
+
+
+@given(
+    x=xps.arrays(dtype=xps.scalar_dtypes(), shape=shared_shapes()),
+    axis=shared_shapes().flatmap(lambda s: st.integers(-len(s) - 1, len(s))),
+)
+def test_expand_dims(x, axis):
+    out = xp.expand_dims(x, axis=axis)
+
+    ph.assert_dtype("expand_dims", x.dtype, out.dtype)
+
+    shape = [side for side in x.shape]
+    index = axis if axis >= 0 else x.ndim + axis + 1
+    shape.insert(index, 1)
+    shape = tuple(shape)
+    ph.assert_result_shape("expand_dims", (x.shape,), out.shape, shape)
+
+    assert_array_ndindex(
+        "expand_dims", x, ah.ndindex(x.shape), out, ah.ndindex(out.shape)
+    )
+
+
+@given(
+    x=xps.arrays(
+        dtype=xps.scalar_dtypes(), shape=hh.shapes(min_side=1).filter(lambda s: 1 in s)
+    ),
+    data=st.data(),
+)
+def test_squeeze(x, data):
```

Comment:
Let's keep the test functions in alphabetical order in the file.

Reply:
This function also has to raise a specific error for invalid axes. Probably just generating all possible axes instead of only the valid ones can prevent the need for `data` here.